Mathlib/Topology/CompletelyRegular.lean

```diff
@@ -0,0 +1,111 @@
+
+## Implementation notes
+
+The present definition `CompletelyRegularSpace` is a slight modification of the one given in
+[russell1974]. There it's assumed that any point `x ∈ Kᶜ` is separated from the closed set `K` by a
+continuous *real* valued function `f` (as opposed to `f` being unit-interval-valued). This can be
+converted to the present definition by replacing a real-valued `f` by `h ∘ g ∘ f`, with
+`g : x ↦ max(x, 0)` and `h : x ↦ min(x, 1)`.
+
+## References
+
+* [Russell C. Walker, *The Stone-Čech Compactification*][russell1974]
+-/
+
+universe u
+
+noncomputable section
+
+open Filter Set Topology
+
+variable {α : Type u} [TopologicalSpace α] [T1Space α]
+variable {γ : Type u} [TopologicalSpace γ] [T2Space γ] [CompactSpace γ]
+
+@[mk_iff completelyRegularSpace_iff]
+class CompletelyRegularSpace (α : Type u) [TopologicalSpace α] [T1Space α] : Prop where
+  completely_regular :
+  ∀ (x : α), ∀ (K : Set α) (_: IsClosed K), Disjoint K {x} →
+    ∃ (f : C(α, ℝ)), EqOn f 0 {x} ∧ EqOn f 1 K ∧ ∀ x, f x ∈ Icc (0 : ℝ) 1
+
+instance CompletelyRegularSpace.t3space [CompletelyRegularSpace α] : T3Space α := by
+  have : T0Space α := by apply T1Space.t0Space
+  have : RegularSpace α := by
+    rw [regularSpace_iff]
+    intros s a hs ha
+    rw [Filter.disjoint_iff]
+    have d : Disjoint s {a} := by
+      exact Eq.mpr (id (propext disjoint_singleton_right ▸ Eq.refl (Disjoint s {a}))) ha
+    let h := CompletelyRegularSpace.completely_regular a s hs d
+    let f := h.choose
+    let ⟨hf, hhf, _⟩ := h.choose_spec
+    let ⟨u, v, hu, hv, hhu, hhv, huv⟩ := T2Space.t2 (0 : ℝ) 1 zero_ne_one
+    have mem_u : f ⁻¹' u ∈ (𝓝 a) := by
+      rw [mem_nhds_iff]
+      use f ⁻¹' u
+      have L : a ∈ ↑f ⁻¹' u := by
+        rw [mem_preimage]
+        have T : f a = 0 := by
+          apply hf
+          rw [mem_singleton_iff]
+        exact Eq.mpr (id (T ▸ Eq.refl (f a ∈ u))) hhu
+      exact ⟨Subset.refl (f ⁻¹' u), IsOpen.preimage f.2 hu, L⟩
+    have mem_v : f ⁻¹' v ∈ (𝓝ˢ s) := by
+      rw [←subset_interior_iff_mem_nhdsSet]
+      have vo := IsOpen.preimage f.2 hv
+      rw [←interior_eq_iff_isOpen] at vo
+      simp only [eqOn_singleton, Pi.zero_apply, ge_iff_le, zero_le_one, not_true, gt_iff_lt,
+        mem_Icc, ContinuousMap.toFun_eq_coe] at vo
+      simp only [eqOn_singleton, Pi.zero_apply, ge_iff_le, zero_le_one, not_true, gt_iff_lt,
+        mem_Icc]
+      rw [vo]
+      intro x hx
+      rw [mem_preimage]
+      have P : (Exists.choose h) x = 1 := hhf hx
+      simp only [eqOn_singleton, Pi.zero_apply, ge_iff_le, zero_le_one, not_true, gt_iff_lt,
+        mem_Icc] at P
+      rw [P]
+      exact hhv
+    have duv : Disjoint (f ⁻¹' v) (f ⁻¹' u) := by
+      rw [Set.disjoint_iff, inter_comm, subset_empty_iff] at huv
+      rw [Set.disjoint_iff, ←preimage_inter, huv, preimage_empty]
+    exact
+      Exists.intro (↑f ⁻¹' v)
+        { left := mem_v, right := Exists.intro (↑f ⁻¹' u) { left := mem_u, right := duv } }
```

Comment:
You might find the anonynous constructor notation helpful here, you can use `\<a, b\>` in place of `Exists.intro a b`, and you can nest this and replace `{left := a, right := b}` with the same thing too.

Reply:
i.e., `⟨f ⁻¹' v, mem_v, f ⁻¹' u, mem_u, duv⟩`